helix-term/src/application.rs

```diff
@@ -49,33 +49,94 @@ pub struct Application {
-        let mut compositor = Compositor::new()?;
-        let size = compositor.size();
 
-        let conf_dir = helix_core::config_dir();
+        // These configuration directories can contain `config.toml` and `languages.toml`.
+        // `local_config_dir` is a `.helix` folder within the projec directory.
+        let config_dir = helix_core::config_dir();
+        let local_config_dir = helix_core::local_config_dir();
+        let cwd = std::env::current_dir().expect("unable to determine current directory");
+        if !config_dir.exists() {
+            std::fs::create_dir_all(&config_dir).ok();
+        }
 
-        let theme_loader =
-            std::sync::Arc::new(theme::Loader::new(&conf_dir, &helix_core::runtime_dir()));
-
-        // load default and user config, and merge both
-        let builtin_err_msg =
-            "Could not parse built-in languages.toml, something must be very wrong";
-        let def_lang_conf: toml::Value =
-            toml::from_slice(include_bytes!("../../languages.toml")).expect(builtin_err_msg);
-        let def_syn_loader_conf: helix_core::syntax::Configuration =
-            def_lang_conf.clone().try_into().expect(builtin_err_msg);
-        let user_lang_conf = std::fs::read(conf_dir.join("languages.toml"))
-            .ok()
-            .map(|raw| toml::from_slice(&raw));
-        let lang_conf = match user_lang_conf {
-            Some(Ok(value)) => Ok(merge_toml_values(def_lang_conf, value)),
-            Some(err @ Err(_)) => err,
-            None => Ok(def_lang_conf),
+        // `config.toml`
+        let mut config = match std::fs::read_to_string(config_dir.join("config.toml")) {
+            Ok(config) => toml::from_str(&config)
+                .map(crate::keymap::merge_keys)
+                .unwrap_or_else(|err| {
+                    eprintln!("Bad config: {}", err);
+                    eprintln!("Press <ENTER> to continue with default config");
+                    use std::io::Read;
+                    // This waits for an enter press.
+                    let _ = std::io::stdin().read(&mut []);
+                    Config::default()
+                }),
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => Config::default(),
+            Err(err) => return Err(Error::new(err)),
         };
 
+        // Config override order: local (cwd) -> local (root) -> global -> default.
+        // Read and parse the `languages.toml` files as TOML objects.
+        let default_lang_config: toml::Value =
+            toml::from_slice(include_bytes!("../../languages.toml"))
+                .expect("failed to read the default `languages.toml`");
+        let lang_config: toml::Value = {
+            let local_cwd_config = load_lang_config(cwd.join(".helix/languages.toml"))?;
+            let local_root_config = load_lang_config(local_config_dir.join("languages.toml"))?;
+            let global_config = load_lang_config(config_dir.join("languages.toml"))?;
```

Comment:
Using strings here feels icky, I think top level constants would be better here for `languages.toml`, `../../languages.toml`, etc.

Reply:
Unfortunately a constant doesn't work for `include_bytes!` because it expects a string literal only.